### ISSUE TITLE
Keyboard-attach footer-bar animation

### ISSUE THREAD

**Issue by Xavierh95** [driftyco/ionic#3537]
Hi,
I have an issue with the keyboard-attach directive on a ion-footer-bar
I'm using Ionic 1.0.0-rc.2 and ionic keyboard plugin.

When I click on an input inside this ion-footer-bar, the keyboard is correctly opened and the footer is on the top of the keyboard.
But during the transition (opening or closing), the ion-footer-bar is not following the keyboard.

I look into the Ionic bundle.js and I see the following code : 
```javascript
IonicModule
.directive('keyboardAttach', function() {
  return function(scope, element, attrs) {
    ionic.on('native.keyboardshow', onShow, window);
    ionic.on('native.keyboardhide', onHide, window);

    //deprecated
    ionic.on('native.showkeyboard', onShow, window);
    ionic.on('native.hidekeyboard', onHide, window);


    var scrollCtrl;

    function onShow(e) {
      if (ionic.Platform.isAndroid() && !ionic.Platform.isFullScreen) {
        return;
      }

      //for testing
      var keyboardHeight = e.keyboardHeight || e.detail.keyboardHeight;
      element.css('bottom', keyboardHeight + "px");
      scrollCtrl = element.controller('$ionicScroll');
      if (scrollCtrl) {
        scrollCtrl.scrollView.__container.style.bottom = keyboardHeight + keyboardAttachGetClientHeight(element[0]) + "px";
      }
    }

   ....
```

The position of the bar is set without animation and the result is really not smooth / nice for the final user.

You can see the result in the following link : https://vimeo.com/125166338

Moreover, as you can see in this video, the scrollview is not scrolled with the opening of the keyboard but after the opening.

Can you help me regarding this issue ?

Best regards,

**Comment by Ionitron** [comment]
Greetings @Xavierh95!

My sensors indicate that you need to update your issue through our [custom issue form](http://ionicframework.com/submit-issue/#/?iid=3537). We are now requiring all issues to be submitted this way, to ensure that we have all of the information necessary to fix them as quickly as possible.

**[Click Here To Update Your Issue](http://ionicframework.com/submit-issue/#/?iid=3537)**

I will have no choice but to close this issue if it is not resubmitted through the form. Please fill out the rest of the form, so that I may use my friendly robot powers to assist you.

Thank you!

<span ionitron-issue-resubmit></span>

**Comment by tlancina** [comment]
I agree, it should animate in.  That being said, it isn't possible to animate perfectly along with the keyboard like a native app, because the native -> JS bridge for the keyboard plugin will execute whenever the main thread has time, so even on iOS where we know the height of the keyboard and the duration of the keyboard animation, we can never start the animation at the same time.

**Comment by Xavierh95** [comment]
I have tried to replace the line 
```javascript
element.css('bottom', keyboardHeight + "px");
```
by something like in (http://forum.ionicframework.com/t/suggestion-animation-for-keyboard-attach-directive/5950)
```javascript
$(element).animate({bottom:keyboardHeight + "px"});
```

But the animation seems to start after the keyboard animation. 

In iOS, we can retrieve the height of the keyboard and the duration, so with an update of the keyboard plugin maybe there is a solution, isn't ?

**Comment by tlancina** [comment]
There is kind of a solution if you are willing to make your app a bit unstable.

evalJS:
https://github.com/driftyco/ionic-plugins-keyboard/blob/master/src/ios/IonicKeyboard.m#L29

evalJSHelper
https://github.com/apache/cordova-ios/blob/master/CordovaLib/Classes/CDVCommandDelegateImpl.m#L83

You can modify the keyboard plugin to use scheduledOnRunLoop:NO
```
[weakSelf.commandDelegate evalJs:[NSString stringWithFormat:@"cordova.plugins.Keyboard.isVisible = true; cordova.fireWindowEvent('native.keyboardshow', { 'keyboardHeight': %@ }); ", [@(keyboardFrame.size.height) stringValue]] scheduledOnRunLoop:NO];
```

However, it might crash occasionally :)

**Comment by Xavierh95** [comment]
Hi, Thanks to your answer !
Actually, I'm an iOS Developer so I have access to xCode. 
I will try this but why it might crash occasionally ?

**Comment by tlancina** [comment]
I seemed to remember experimenting with it before and assuming the crashes I experienced were due to the change.  However, XCode crashes for me all the time during debugging with Safari, so that's probably a bad assumption to make!

Taking a closer look, it looks like the only reason it would crash is if you called evalJS from somewhere other than the main thread, which isn't case for the keyboard plugin.

That being said, it does appear that you could actually eval the Cordova dispatchEvent in the middle of another function's execution, which might be weird.

Unfortunately I'm not familiar enough with iOS to definitively say this is safe :) I just assumed Cordova implemented evalJS for plugins for a reason, otherwise you could probably just use UIWebView's stringByEvaluatingJavaScriptFromString directly I would think.

**Comment by tlancina** [comment]
@Xavierh95 here's the [response from some of the Cordova team](https://mail-archives.apache.org/mod_mbox/cordova-dev/201504.mbox/%3CCABiQX1XfbXdBxf0k%2BcJKJb7Snc-xLYi95h7SHdKVi-EbmNZHTA%40mail.gmail.com%3E).  I'll still add an animation to the keyboard-attach directive, but I probably won't change the keyboard plugin since even firing right away the CSS animation is not in sync.

**Comment by lisovet** [comment]
Any updates on this @tlancina?

**Comment by allucardster** [comment]
Any updates about this issue?

**Comment by prabhu2003ece** [comment]
Any updates?

**Comment by tlancina** [comment]
As far as I know, there is no way to perfectly sync up an HTML input with the native keyboard animation. However, you might want to take a look at https://github.com/EddyVerbruggen/cordova-plugin-native-keyboard, as a native input can animate with the keyboard and if you're building a chat-style application.

**Comment by prabhu2003ece** [comment]
I tried a similar approach as @Xavierh95 by modifying the keyboardAttach directive. 
I was able to get a near native experience. Again, this is not like a actual native experience. But, atleast looks better than the original.

      setTimeout(function () {
    	   scope.$apply(function() {
    		   element.css('bottom', keyboardHeight + "px");
        	   scrollCtrl = element.controller('$ionicScroll');
        	      if (scrollCtrl) {
        	        scrollCtrl.scrollView.__container.style.bottom = keyboardHeight + keyboardAttachGetClientHeight(element[0]) + "px";
        	    }
    	   });
    	}, 50);

**Comment by tskweres** [comment]
any updates? Not animating the bottom chat bar for a chat app is pretty killer

**Comment by geshub** [comment]
It would be great to find a workaround for those working on chat apps
